### PR TITLE
Index slice

### LINKED ISSUE #300: IndexSlice typing is no longer correct
**Describe the bug**
Prior to #282, `pd.IndexSlice` returned a tuple with specific types.  That is no longer the case.

**To Reproduce**
```python
import pandas as pd

df = (
    pd.DataFrame({"x": [1, 2, 2, 3, 4], "y": [10, 20, 30, 40, 10]})
    .assign(z=lambda df: df.x * df.y)
    .set_index(["x", "y"])
)
ind = pd.Index([2, 3])
reveal_type(pd.IndexSlice[ind, :])
```
With pandas-stubs 1.4.4.220911, we get:

pyright reports:
```text
isli.py:4:13 - information: Type of "pd.IndexSlice[ind, :]" is "tuple[Unknown, ...]"
```
mypy reports:
```text
isli.py:4: note: Revealed type is "builtins.tuple[Any, ...]"
```

With pandas-stubs 1.4.4.220906, we get:

pyright:
```text
isli.py:9:13 - information: Type of "pd.IndexSlice[ind, :]" is "tuple[str | str_ | bytes | date | datetime | timedelta | bool | complex | Timestamp | Timedelta | slice, ...]"
```

mypy:
```text
isli.py:9: note: Revealed type is "builtins.tuple[Union[Union[builtins.str, numpy.str_], Union[builtins.str, builtins.bytes, datetime.date, datetime.datetime, datetime.timedelta, builtins.bool, builtins.complex, pandas._libs.tslibs.timestamps.Timestamp, pandas._libs.tslibs.timedeltas.Timedelta], builtins.slice], ...]"
```

**Please complete the following information:**
 - OS: Windows 10
 - python version:  3.10
 - mypy 0.971, pyright 1.269
 - version of installed `pandas-stubs`:  1.4.4.220911

**Additional context**

@bashtage needs to look at this, because change in #282 is what messed this up.

### PR DESCRIPTION
**Pull request by Dr-Irv**
- [x] Closes #300
- [x] Tests added
    - `test_frame.py:test_indexslice_getitem()`

### CODE REVIEW

File: pandas-stubs/core/indexing.pyi [hunk 1]
```diff
@@ -1,22 +1,20 @@
...
 
-_IndexSliceT = TypeVar("_IndexSliceT", bound=Union[StrLike, Scalar, slice])
+_IndexSliceTuple = slice | tuple[Index | MaskType | Scalar | list[Scalar] | slice, ...]
+
+_IndexSliceTupleT = TypeVar("_IndexSliceTupleT", bound=_IndexSliceTuple)
 
-class _IndexSlice(Generic[_IndexSliceT]):
-    def __getitem__(self, arg) -> tuple[_IndexSliceT, ...]: ...
+class _IndexSlice:
+    def __getitem__(self, arg: _IndexSliceTupleT) -> _IndexSliceTupleT: ...
```
Comment (bashtage) [inline pandas-stubs/core/indexing.pyi:17]: Isn't it always a tuple, even when slicing 1d?  In my experiments I always say some soft of tuple whether slicing 1 level 2 or 3.

Comment (bashtage) [reply pandas-stubs/core/indexing.pyi:17]: Or is slice a subclass of tuple somehow?

File: tests/test_frame.py [hunk 2]
```diff
@@ -1278,6 +1278,18 @@ def test_indexslice_setitem():
...
+def test_indexslice_getitem():
+    # GH 300
+    df = (
+        pd.DataFrame({"x": [1, 2, 2, 3, 4], "y": [10, 20, 30, 40, 10]})
+        .assign(z=lambda df: df.x * df.y)
+        .set_index(["x", "y"])
+    )
+    ind = pd.Index([2, 3])
+    check(assert_type(pd.IndexSlice[ind, :], tuple[pd.Index, slice]), tuple)
+    check(assert_type(df.loc[pd.IndexSlice[ind, :]], pd.DataFrame), pd.DataFrame)
```
Comment (bashtage) [inline tests/test_frame.py:1290]: Ideally you would have a test that would show all possible types can be generated.

**Review by bashtage — CHANGES_REQUESTED** [review]
Some questions and a test that shows all types allowed in the typevar can be used.

File: pandas-stubs/core/indexing.pyi (hunk 1 above)
Comment (Dr-Irv) [reply pandas-stubs/core/indexing.pyi:22]: `_IndexSliceTuple` is either a `slice` or a `tuple`.  So if you do `pd.IndexSlice[1:2]`, that returns a `slice` not a `tuple`

File: tests/test_frame.py (hunk 2 above)
Comment (Dr-Irv) [reply tests/test_frame.py:1290]: Some of them are tested elsewhere, but adding more tests is a good idea.  Will work on that.

File: pandas-stubs/core/indexing.pyi (hunk 1 above)
Comment (twoertwein) [reply pandas-stubs/core/indexing.pyi:22]: I must admit I have never directly used `pd.IndexSlice.__getitem__` - it seems to simply be an identity function.

```py
>>> obj = object()
>>> pd.IndexSlice[obj] is obj
True
```

Comment (Dr-Irv) [reply pandas-stubs/core/indexing.pyi:22]: If you are doing slicing on a `MultiIndex`), it really becomes necessary.  See the docs at https://pandas.pydata.org/docs/reference/api/pandas.IndexSlice.html?highlight=indexslice#pandas.IndexSlice for a good example.

File: tests/test_frame.py (hunk 2 above)
Comment (Dr-Irv) [reply tests/test_frame.py:1290]: I added a bunch of tests in the latest commit.

**Comment by Dr-Irv** [comment]
@bashtage @twoertwein this is all green now, so should be good to go

**Review by twoertwein — APPROVED** [review]
LGTM but I'm not familiar with IndexSlice

**Comment by twoertwein** [comment]
Thanks @Dr-Irv

**Pull request merged**